plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -647,32 +655,51 @@ public Iterator<TableColumnsMetadata> streamTableColumns(ConnectorSession sessio
-                            return Stream.of(TableColumnsMetadata.forRedirectedTable(tableName));
-                        }
 
-                        Table icebergTable = catalog.loadTable(session, tableName);
-                        List<ColumnMetadata> columns = getColumnMetadatas(icebergTable.schema());
-                        return Stream.of(TableColumnsMetadata.forTable(tableName, columns));
-                    }
-                    catch (TableNotFoundException e) {
-                        // Table disappeared during listing operation
-                        return Stream.empty();
-                    }
-                    catch (UnknownTableTypeException e) {
-                        // Skip unsupported table type in case that the table redirects are not enabled
-                        return Stream.empty();
-                    }
-                    catch (RuntimeException e) {
-                        // Table can be being removed and this may cause all sorts of exceptions. Log, because we're catching broadly.
-                        log.warn(e, "Failed to access metadata of table %s during streaming table columns for %s", tableName, prefix);
-                        return Stream.empty();
+        BlockingQueue<TableColumnsMetadata> columnQueue = new LinkedBlockingQueue<>();
+        List<Future<Void>> tasks = new ArrayList<>();
+        for (SchemaTableName tableName : schemaTableNames) {
+            tasks.add(CompletableFuture.runAsync(() -> {
+                try {
+                    if (redirectTable(session, tableName).isPresent()) {
+                        columnQueue.add(TableColumnsMetadata.forRedirectedTable(tableName));
                     }
-                })
-                .iterator();
+
+                    Table icebergTable = catalog.loadTable(session, tableName);
+                    List<ColumnMetadata> columns = getColumnMetadatas(icebergTable.schema());
+                    columnQueue.add(TableColumnsMetadata.forTable(tableName, columns));
+                }
+                catch (TableNotFoundException e) {
+                    // Table disappeared during listing operation
+                }
+                catch (UnknownTableTypeException e) {
+                    // Skip unsupported table type in case that the table redirects are not enabled
+                }
+                catch (RuntimeException e) {
+                    // Table can be being removed and this may cause all sorts of exceptions. Log, because we're catching broadly.
+                    log.warn(e, "Failed to access metadata of table %s during streaming table columns for %s", tableName, prefix);
+                }
+            }, metadataListingExecutor));
+        }
+
+        return new Iterator<>() {
+            @Override
+            public boolean hasNext()
+            {
+                return !columnQueue.isEmpty() || !tasks.stream().allMatch(Future::isDone);
```

Comment:
Checking all tasks for completed futures is very expensive. This should be an O(1) operation to pull completed tasks.